File: vripper-server/src/main/java/tn/mnlr/vripper/services/DownloadSpeedService.java
package tn.mnlr.vripper.services;

import lombok.Getter;
import org.springframework.scheduling.annotation.EnableScheduling;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import tn.mnlr.vripper.event.Event;
import tn.mnlr.vripper.event.EventBus;

import java.util.concurrent.atomic.AtomicLong;

@Service
@EnableScheduling
public class DownloadSpeedService {

  private final AtomicLong read = new AtomicLong(0);
  private final EventBus eventBus;
  @Getter private long currentValue;
  private boolean allowWrite = false;

  public DownloadSpeedService(EventBus eventBus) {
    this.eventBus = eventBus;
  }

  public void increase(long read) {
    if (allowWrite) {
      this.read.addAndGet(read);
    }
  }

  @Scheduled(fixedDelay = 1000)
  private void calc() {
    allowWrite = false;
    long newValue = read.getAndSet(0);
    if (newValue != currentValue) {
      currentValue = newValue;
      eventBus.publishEvent(Event.wrap(Event.Kind.BYTES_PER_SECOND, currentValue));
    }
    allowWrite = true;
  }
}

File: vripper-core/src/main/kotlin/me/mnlr/vripper/services/DownloadSpeedService.kt
package me.mnlr.vripper.services

import kotlinx.coroutines.*
import me.mnlr.vripper.event.DownloadSpeedEvent
import me.mnlr.vripper.event.EventBus
import me.mnlr.vripper.event.QueueStateEvent
import me.mnlr.vripper.formatSI
import me.mnlr.vripper.model.DownloadSpeed
import java.util.concurrent.atomic.AtomicLong

class DownloadSpeedService(
    private val eventBus: EventBus,
) {

    companion object {
        const val DOWNLOAD_POLL_RATE = 2500
    }

    private val coroutineScope = CoroutineScope(Dispatchers.Default)
    private val bytesCount = AtomicLong(0)
    private var job: Job? = null

    fun init() {
        coroutineScope.launch {
            eventBus.subscribe<QueueStateEvent> {
                if (it.queueState.running + it.queueState.remaining > 0) {
                    if (job == null || job?.isActive == false) {
                        job = coroutineScope.launch {
                            eventBus.publishEvent(DownloadSpeedEvent(DownloadSpeed(0L.formatSI())))
                            while (isActive) {
                                delay(DOWNLOAD_POLL_RATE.toLong())
                                val newValue = bytesCount.getAndSet(0)

                                eventBus.publishEvent(DownloadSpeedEvent(DownloadSpeed(((newValue * 1000) / DOWNLOAD_POLL_RATE).formatSI())))

                            }
                        }
                    }
                } else {
                    job?.cancel()
                    coroutineScope.launch {
                        delay(DOWNLOAD_POLL_RATE + 500L)
                        eventBus.publishEvent(DownloadSpeedEvent(DownloadSpeed(0L.formatSI())))
                    }
                }
            }
        }
    }

    fun reportDownloadedBytes(count: Long) {
        bytesCount.addAndGet(count)
    }
}
